mongo/options/clientoptions.go

```diff
@@ -279,26 +279,32 @@ func (c *ClientOptions) ApplyURI(uri string) *ClientOptions {
 			tlsConfig.InsecureSkipVerify = true
 		}
 
+		var x509Subject string
+		var keyPasswd string
+		if cs.SSLClientCertificateKeyPasswordSet && cs.SSLClientCertificateKeyPassword != nil {
+			keyPasswd = cs.SSLClientCertificateKeyPassword()
+		}
 		if cs.SSLClientCertificateKeyFileSet {
-			var keyPasswd string
-			if cs.SSLClientCertificateKeyPasswordSet && cs.SSLClientCertificateKeyPassword != nil {
-				keyPasswd = cs.SSLClientCertificateKeyPassword()
-			}
-			s, err := addClientCertFromFile(tlsConfig, cs.SSLClientCertificateKeyFile, keyPasswd)
-			if err != nil {
-				c.err = err
-				return c
-			}
+			x509Subject, err = addClientCertFromConcatenatedFile(tlsConfig, cs.SSLClientCertificateKeyFile, keyPasswd)
+		} else if cs.SSLCertificateFileSet || cs.SSLPrivateKeyFileSet {
+			x509Subject, err = addClientCertFromSeparateFiles(tlsConfig, cs.SSLCertificateFile,
+				cs.SSLPrivateKeyFile, keyPasswd)
+		}
+		if err != nil {
+			c.err = err
+			return c
+		}
 
-			// If a username wasn't specified, add one from the certificate.
-			if c.Auth != nil && strings.ToLower(c.Auth.AuthMechanism) == "mongodb-x509" && c.Auth.Username == "" {
-				// The Go x509 package gives the subject with the pairs in reverse order that we want.
-				pairs := strings.Split(s, ",")
-				for left, right := 0, len(pairs)-1; left < right; left, right = left+1, right-1 {
-					pairs[left], pairs[right] = pairs[right], pairs[left]
-				}
-				c.Auth.Username = strings.Join(pairs, ",")
+		// If a username wasn't specified and a certificate was provided, add one from the certificate.
+		if x509Subject != "" && c.Auth != nil && strings.ToLower(c.Auth.AuthMechanism) == "mongodb-x509" &&
```

Comment:
I don't think this check is necessary.  If X509 is selected, then a cert must have been provided.